runtime/compiler/env/DependencyTable.cpp

```diff
@@ -0,0 +1,198 @@
+      {
+      deactivateTable();
+      }
+   }
+
+OffsetEntry *
+TR_AOTDependencyTable::getOffsetEntry(uintptr_t offset, bool create)
+   {
+   auto it = _offsetMap.find(offset);
+   if (it != _offsetMap.end())
+      return &it->second;
+
+   if (create)
+      {
+      PersistentUnorderedSet<J9Class *> loadedClasses(PersistentUnorderedSet<J9Class *>::allocator_type(TR::Compiler->persistentAllocator()));
+      return &(*_offsetMap.insert(it, {offset, {loadedClasses}})).second;
+      }
+
+   return NULL;
+   }
+
+void
+TR_AOTDependencyTable::invalidateUnloadedClass(TR_OpaqueClassBlock *clazz)
+   {
+   uintptr_t classOffset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+   if (!_sharedCache->isClassInSharedCache(clazz, &classOffset))
+      return;
+
+
+   OMR::CriticalSection cs(_tableMonitor);
+   if (!isActive())
+      return;
+
+   auto entry = getOffsetEntry(classOffset, false);
+   if (entry)
+      {
+      entry->_loadedClasses.erase((J9Class *)clazz);
+      if (entry->_loadedClasses.empty())
+         _offsetMap.erase(classOffset);
+      }
+   }
+
+// If an entry exists for a class, remove it. Otherwise, if we should
+// revalidate, add an entry if the class has a valid chain.
+void
+TR_AOTDependencyTable::recheckSubclass(J9Class *ramClass, uintptr_t offset, bool shouldRevalidate)
+   {
+   auto it = _offsetMap.find(offset);
+   if (it != _offsetMap.end())
+      {
+      auto c_it = it->second._loadedClasses.find(ramClass);
+      if (c_it != it->second._loadedClasses.end())
+         it->second._loadedClasses.erase(c_it);
+      else if (_sharedCache->classMatchesCachedVersion(ramClass, NULL))
+         it->second._loadedClasses.insert(c_it, ramClass);
+      }
+   else if (shouldRevalidate && _sharedCache->classMatchesCachedVersion(ramClass, NULL))
+      {
+      PersistentUnorderedSet<J9Class *> loadedClasses(PersistentUnorderedSet<J9Class *>::allocator_type(TR::Compiler->persistentAllocator()));
+      loadedClasses.insert(ramClass);
+      _offsetMap.insert(it, {offset, {loadedClasses}});
+      }
+   }
+
+// In a class redefinition event, an old class is replaced by a fresh class. If
+// the ROM class offset changed as a result, it and all its subclasses that
+// formerly had valid chains will now be guaranteed not to match, so the entries
+// for these must be removed. If the new offset is valid, any class that didn't
+// have an entry should be rechecked.
+void
+TR_AOTDependencyTable::invalidateRedefinedClass(TR_PersistentCHTable *table, TR_J9VMBase *fej9, TR_OpaqueClassBlock *oldClass, TR_OpaqueClassBlock *freshClass)
+   {
+   if (!isActive())
+      return;
+
+   uintptr_t freshClassOffset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+   uintptr_t oldClassOffset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+   if (!_sharedCache->isClassInSharedCache(freshClass, &freshClassOffset) && !_sharedCache->isClassInSharedCache(oldClass, &oldClassOffset))
+      return;
+
+   if (oldClassOffset == freshClassOffset)
```

Comment:
If the offsets are the same, do you need to add `oldClass` into the table? I guess if something needs the class chain validation status, it will use the new class, but is there anything that might still use the old class to look up the status? Even something does, does it matter if it can't find it? I'm gonna guess the answer is no, but I figure it's worth pointing out just in case.

Reply:
The `oldClass` would have been tracked already in `classLoadEvent`. I actually need to invalidate the `oldClass` and then register that `freshClass` was loaded. Should be fixed.